src/V3HierBlock.cpp

```diff
@@ -0,0 +1,420 @@
+}
+
+void V3HierBlock::writeCommandFile(bool forCMake) const {
+    vl_unique_ptr<std::ofstream> of(V3File::new_ofstream(commandFileName(forCMake)));
+    *of << "--cc\n";
+
+    if (!forCMake) {
+        for (V3HierBlock::HierBlockSet::const_iterator child = m_children.begin();
+             child != m_children.end(); ++child) {
+            *of << v3Global.opt.makeDir() << "/" << (*child)->hierWrapper(true) << "\n";
+        }
+    }
+    *of << "-Mdir " << v3Global.opt.makeDir() << "/" << hierPrefix() << " \n";
+    writeCommonInputs(of.get(), forCMake);
+    const V3StringList& commandOpts = commandOptions(false);
+    for (V3StringList::const_iterator it = commandOpts.begin(); it != commandOpts.end(); ++it) {
+        *of << (*it) << "\n";
+    }
+    *of << hierBlockOptions().front() << "\n";
+    for (HierBlockSet::const_iterator child = m_children.begin(); child != m_children.end();
+         ++child) {
+        *of << hierBlockOptions().front() << "\n";
+    }
+    *of << v3Global.opt.allArgsStringForHierBlock(false) << "\n";
+}
+
+string V3HierBlock::commandFileName(bool forCMake) const {
+    return hierCommandFileName(hierPrefix(), forCMake);
+}
+
+//######################################################################
+// Collect how hierarchy blocks are used
+class HierBlockUsageCollectVisitor : public AstNVisitor {
+    // NODE STATE
+    // AstNode::user1()            -> bool. Processed
+    AstUser1InUse m_inuser1;
+
+    // STATE
+    typedef std::set<const AstModule*> ModuleSet;
+    V3HierBlockPlan* const m_planp;
+    AstModule* m_modp;
+    ModuleSet m_referred;  // Modules that have hier_block pragma
+    V3HierBlock::GParams m_gparams;  // list of variables that is AstVarType::GPARAM
+
+    virtual void visit(AstModule* nodep) VL_OVERRIDE {
+        // Don't visit twice
+        if (nodep->user1SetOnce()) return;
+        UINFO(5, "Checking " << nodep->prettyNameQ() << " from "
+                             << (m_modp ? m_modp->prettyNameQ() : string("null")) << std::endl);
+        AstModule* const prevModulep = m_modp;
+        ModuleSet prevReferred;
+        V3HierBlock::GParams prevGParams;
+        if (nodep->hierBlock()) {
+            m_modp = nodep;
+            prevReferred.swap(m_referred);
+        }
+        prevGParams.swap(m_gparams);
+
+        iterateChildren(nodep);
+
+        if (nodep->hierBlock()) {
+            m_planp->add(nodep, m_gparams);
+            for (ModuleSet::const_iterator it = m_referred.begin(); it != m_referred.end(); ++it) {
+                m_planp->registerUsage(nodep, *it);
+            }
+            m_modp = prevModulep;
+            m_referred.swap(prevReferred);
+        }
+        m_gparams.swap(prevGParams);
+    }
+    virtual void visit(AstCell* nodep) VL_OVERRIDE {
+        // Visit used module here to know that the module is hier_block or not.
+        // This visitor behaves almost depth first search
+        if (AstModule* modp = VN_CAST(nodep->modp(), Module)) {
```

Comment:
What about interfaces?  Check interfaces work in some tests if didn't already.

Reply:
Actually the interface from AstCell should be ignored in the function.
But thanks to your comment here, I noticed that modport cannot be used at the hierarchical block boundary as Verilator does not support interfaced port on top level module.
(using inside hierarchy block is fine.)

Add a check and test in 13d626e33f379c5ff1ae79098259cb45f5e85ed7.
- using interface inside hierarchical block works
- show proper error message if modport is used at the hierarchical block top